### ISSUE TITLE
Try 2D nnUnet

### ISSUE THREAD

**Issue by jcohenadad** [ivadomed/model_seg_mouse-sc_wm-gm_t1#30]
Given the high performance of the nnUnet in general, it would be a good idea to have a benchmark with this architecture.

Todo (update if necessary):

- [x] Write a script to convert the current BIDS data to a nnUnet architecture. See https://github.com/ivadomed/data-conversion
- [x] Look into [MONAI's nnUnetv2 implementation](https://github.com/Project-MONAI/MONAI/tree/4bfc8e97661d44cc13553721e5025005757bad91/monai/apps/nnunet)
- [x] Write training script with nnUnet that uses wanbd. Ideally, the code would be merged with the existing [training](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/main/train.py) and [testing](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/main/test.py) codes 

Related to #17

Tagging @naga-karthik @louisfb01 @valosekj for help

**Comment by plbenveniste** [comment]
Hi !
It seems that the dataset that I have received is not complete : is it normal that the participants.tsv file only contains : 
participant_id	species
sub-mouse1	mouse
sub-mouse2	mouse
sub-mouse3	mouse

@naga-karthik  @valosekj

**Comment by valosekj** [comment]
Hi @plbenveniste!

>  Indeed, it seems that the current "create_msd_json_from_bids.py" file doesn't work in the case where there are two possible suffixes. 

`create_msd_json_from_bids.py` is the old script compatible with nnUNetv1. We are currently developing a new [convert_bids_to_nnUnetv2.py](https://github.com/ivadomed/data-conversion/pull/11) script compatible with nnUnetv2.

Regarding two suffixes; this is exactly what we are working on in [another project](https://github.com/ivadomed/model_seg_sci). I guess @naga-karthik has an updated version of the conversion script compatible with two suffixes.

**Comment by plbenveniste** [comment]
Thank you for your answer. In my question I meant two '--label-suffix' so I think it's equivalent to what you are working on. The goal in my case is to have one label for grey matter and one label for white matter. 
I'll look into it and tell you if I come up with a working solution. 
However, do you have any useful resources which I could use to move further ?

**Comment by naga-karthik** [comment]
Hey @plbenveniste, thanks for your questions? Let us go through them step-by-step: 

> I am trying to convert the dataset from BIDS to nnUnet.

Unless I missed something, shouldn't you be using the `convert_bids_to_nnunet.py` to (presumably) train an nnUNet model? The file `create_msd_json_from_bids.py ` is specifically for training MONAI models and hence the image, label pairings have more options 

> it seems that the current "create_msd_json_from_bids.py" file doesn't work in the case where there are two possible suffixes.

yes, you're right. In most cases, it assumes that there are two or more "contrasts" but a single (common) label. 

> With the zurich-mouse dataset, we have : _label-GM_mask and _label-WM_mask.

A quick workaround for this would be to slightly modify the `group_by_contrasts` if condition [here](https://github.com/ivadomed/data-conversion/blob/main/dataset_conversion/create_msd_json_from_bids.py#L241-L242) to have two labels instead of two contrasts (i.e. label_0000 for GM_mask and label_0001 for WM_mask). This is assuming that your dataset does not have multiple sessions (if it does, then you'd have to modify the if condition for `group_by_sessions` instead)

> However, do you have any useful resources which I could use to move further ?

If the above does not work, then you could check out the [dataset conversion instructions](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/dataset_format.md) on nnUNet's repo. Even [monai's tutorials](https://github.com/Project-MONAI/tutorials) have some examples on dataset conversion. 

Hope this helps!

**Comment by plbenveniste** [comment]
Indeed, I was mistaken. I thought I had to use the `create_msd_json_from_bids.py` to create the `dict_split.json` file. I ended up creating it by hand. 
Thank you for your help. 
I am now trying to figure out how to use two labels in the case of `convert_bids_to_nnunet.py`.

**Comment by jcohenadad** [comment]
@plbenveniste Can you please move [your developments](https://github.com/plbenveniste/nnunet_seg_mouse-sc_gm_wm-t1) under a branch in this repository? It will be easier to manage/crossref discussions/issues/code/results if development stays in one repos. Thx!

**Comment by plbenveniste** [comment]
Done. 
I called the branch `plb/nnunet`

**Comment by jcohenadad** [comment]
> Done. I called the branch plb/nnunet

great! could you also open a PR so we can discuss developments / code design in the PR directly

**Comment by plbenveniste** [comment]
Update : 
- removed the notebook (which was useless)
- updated the [requirements.txt](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/plb/nnunet/requirements.txt) file
- modified the [conversion file](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/plb/nnunet/convert_bids_to_nnunet.py) to take every image and use those that are labeled in the training and those that aren't for inference. (it doesn't require a split_dict file anymore)
- removed the split_dict file
- uploaded everything on romane (waiting for a gpu to be free)

To do: 
- update ReadMe file
- run a training with 1 fold (I have to think of how relevant it is to use more)
- run inference
- interpret results

**Comment by plbenveniste** [comment]
<img width="647" alt="Screen Shot 2023-06-01 at 11 17 39 AM" src="https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/assets/67429280/75ad8d01-935d-4ebe-8906-59d5dcc639fc">
 
Training crashed at epoch 74. No information in the [training_log_1.txt](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/files/11626940/training_log_1.txt) regarding the reasons of the crash 
Observation: it seems that the model is converging towards a low pseudo-dice value ~0.2

**Comment by plbenveniste** [comment]
Training and testing with the 2D nnU-Net. 
The nnU-Net performs around 0,8 in terms of pseudo-dice. 
But every prediction outputs nothing. This makes sense since more than 80% of the dataset is not annotated. 

Solution: 
- extracting every labeled slice in order to retrain the 2d model on those slices only ? 
- use multiple 3d nnU-Net prediction to get slices annotated at different levels of the spinal cord and aggregate them afterwards ?

**Comment by plbenveniste** [comment]
I tried using multiple runs of a 3D nnU-Net, in order to get different slices annotated and concatenate them in order to end-up with a full annotation of the spinal cord. It didn't work as the model always outputs the annotations for the same slices.

**Comment by jcohenadad** [comment]
> use multiple 3d nnU-Net prediction to get slices annotated at different levels of the spinal cord and aggregate them afterwards ?

no, this is not the right approach. A 3D UNet only makes sense if there is spatial autocorrelation across th 3rd dimension, which is not the case if you concatenate slices that are physically far apart. 

I would just do 2D nnUnet, not 3D. And in fact, I noticed that the title of the issue said "3D nnUnet", but it was a mistake, i meant 2D nnUnet.

**Comment by plbenveniste** [comment]
UPDATE:
I built a script to transform the dataset and extract every labelled slice. It outputs the following files: a nifti file for the slice of the MR image and 2 nifti files for the mask (one for WM and one for GM).  The dataset folders are built to have the same organisation (so that it works for the [convert_bids_to_nnunet.py](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/plb/nnunet/convert_bids_to_nnunet.py) file.

TO DO: 
- train the model on this dataset
- test if if works better (I hope that the model can accept nifti files of different sizes because it is trained on (200,200,1) but tested on (200,200,500)
- if above doesn't work: convert each nifti file of shape (200,200,500) into 500 nifti files (or the other way around)

**Comment by plbenveniste** [comment]
UPDATE: 
- Modification of the files [extract_slices.py](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/plb/nnunet/extract_slices.py) and [convert_bids_to_nnunet.py](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/plb/nnunet/convert_bids_to_nnunet.py) to fix small bugs. 
- Edited the [README.md](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/plb/nnunet/README.md) file
- Currently running training on a 2D nnU-Net

TO DO: 
- Try inference (and see if the dimension difference between training files and testing files causes a problem)

**Comment by plbenveniste** [comment]
**Training results:**
<img width="652" alt="image" src="https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/assets/67429280/8fe2ddd6-3d71-4500-b27e-14a31a9695a1">

Pseudo-dice on validation set: 0.89

**Testing results**
![image](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/assets/67429280/c0d89c82-04af-4281-8815-86a3836089ac)
Fully labelled nifti files: how to determine if correct ? 
Some bugs among the labels: 
![image](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/assets/67429280/3c60f20f-5394-4180-8cc0-f93afdb24a6a)
